### PR TITLE
Add SLAM 2d map to prime-blocks

### PR DESCRIPTION
**Pull request by micheal-parks**
This PR transplants the 2d Slam map (currently in the RC) to prime-blocks with small tweaks that makes the API less opinionated:

```svelte
<SlamMap2d
  pointcloud={new Uint8Array()} // An arraybuffer representing a PCD file.
  basePose={{ x: number, y: number, z: number }} // An optional pose of a robot.
  destination={{ x: number, y: number }} // An optional user-specificed robot destination.
  helpers={true | false} // Whether or not scene helpers should be rendered. Default true.
  motionPath={'0.1,0.2\n0.3,0.4\n'} // An optional motion path.
/>
``` 

slammy

The map was mostly moved as-is to help unblock Steven who will need it in app soon.

It also bundles in the following fixes / improvements for expediency:
* Improve types for bound input HTML elements
* Remove flex class from icon elements

### CODE REVIEW

File: packages/blocks/.eslintrc.cjs [hunk 1]
```diff
@@ -12,6 +12,15 @@ module.exports = {
...
+    // @todo(mp APP-2482): move to js-config
     'sonarjs/no-duplicate-string': 'off',
+    'no-undef-init': 'off',
+    'unicorn/no-useless-undefined': 'off',
+
+    // @todo(mp APP-2482): diagnose why these aren't working correctly
+    '@typescript-eslint/no-unsafe-assignment': 'off',
+    '@typescript-eslint/no-unsafe-return': 'off',
+    '@typescript-eslint/no-unsafe-member-access': 'off',
+    '@typescript-eslint/no-unsafe-call': 'off',
```
Comment (micheal-parks) [inline packages/blocks/.eslintrc.cjs:24]: Our TS eslint plugin is not able to type check svelte components. Added a ticket to investigate [here](https://viam.atlassian.net/browse/APP-2482) and turned off for now since svelte-check will still catch these since we're in TS strict mode.

File: packages/blocks/src/lib/assets/images/base-marker.txt
Comment (mcous) [inline packages/blocks/src/lib/assets/images/base-marker.txt:1]: Why is this file needed? I thought Vite is capable of importing `png` as a data URL

File: packages/blocks/src/lib/assets/images/converter.js
Comment (mcous) [inline packages/blocks/src/lib/assets/images/converter.js:1]: Same comment about committing the image data url; it seems a little prone to desync. I think it'd be better to import `png` directly, or - if we're not comfortable with the whims of `assetInlineLimit` - add this logic as a [Vite import plugin](https://stackoverflow.com/questions/73847316/import-raw-image-data-into-a-script-with-vite), instead

**Review by mcous — COMMENTED** [review]
Quick thought on data URLs

File: packages/blocks/.eslintrc.cjs (hunk 1 above)
Comment (mcous) [reply packages/blocks/.eslintrc.cjs:24]: > out TS eslint plugin is not able to type check svelte components

Weird that it only seems to be happening with the `on:create` arguments of various `T.[]` components. If this was a general problem I'd expect to see a bunch more errors, both here and in app, which uses the same ESLint config.

Perhaps it's safer to use ignore comments rather than disable the rules wholesale, since it seems to be limited to `@threlte/core`?

File: packages/blocks/src/lib/slam-map-2d/points.svelte [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+extend({ MeshDiscardMaterial });
+
+/** A buffer representing a .pcd file */
+export let pointcloud: Uint8Array | undefined;
+
+/** The size of each individual point */
+export let size: number;
+
+type $$Events = {
```
Comment (ethanlook) [inline packages/blocks/src/lib/slam-map-2d/points.svelte:25]: Why `$$Events` here and `Events` in `packages/blocks/src/lib/slam-map-2d/index.svelte`?

Does declaring the type separately show up in IDE suggestions as expected?

File: packages/blocks/tsconfig.json [hunk 3]
```diff
@@ -5,6 +5,7 @@
   ],
   "compilerOptions": {
     "types": ["svelte", "vite/client", "vitest"],
+    "noPropertyAccessFromIndexSignature": false,
```
Comment (ethanlook) [inline packages/blocks/tsconfig.json:8]: Do we have a ticket to track moving this into `js-config`? Do we want that in `js-config`?

**Review by ethanlook — COMMENTED** [review]
There are several unrelated things in this PR - would it be possible to split this into smaller PRs? In general, I strongly prefer small PRs and don't love the pattern of slipping things into unrelated PRs. Happy to chat more offline.

**Comment by micheal-parks** [comment]
> would it be possible to split this into smaller PRs?

@ethanlook I'd normally agree - the only reason I'm bundling like this is because of deliverable pressure coming from the team awareness side. Once obstacle avoidance is wrapped up I'm happy to take the time to split things - and this is my normal modus operandi - up but I'm honestly too exhausted right now to take the time 😞

**Comment by ethanlook** [comment]
> > would it be possible to split this into smaller PRs?
> 
> @ethanlook I'd normally agree...

I hear you and I totally get it! I appreciate the acknowledgment! Carry on 🚀

**Review by ethanlook — APPROVED** [review]

File: packages/blocks/tsconfig.json (hunk 3 above)
Comment (micheal-parks) [reply packages/blocks/tsconfig.json:8]: We do have one! Ticket here: https://viam.atlassian.net/browse/APP-2482.

I think the best course of action is to disable until we can fix at the source

File: packages/blocks/src/lib/slam-map-2d/points.svelte (hunk 2 above)
Comment (micheal-parks) [reply packages/blocks/src/lib/slam-map-2d/points.svelte:25]: `$$Events`` has a different effect: it singles complete changes to event dispatcher payload rather than the contents of a custom event's detail object. It's used within threlte applications because of the usage of `createRawEventDispatcher`, which bypasses using CustomEvent objects since we're working in webgl rather than HTML-land. All intellisense is preserved 👍

File: packages/blocks/src/lib/assets/images/base-marker.txt
Comment (micheal-parks) [reply packages/blocks/src/lib/assets/images/base-marker.txt:1]: Yeah I'm also very not happy with this setup, I'm only preserving it at the moment because direct png imports failed with Vite when we initially set this up in RC + imported into app. I'd rather preserve the thing that we know works while the deadlines are tight and do the good thing afterwards

File: packages/blocks/.eslintrc.cjs (hunk 1 above)
Comment (micheal-parks) [reply packages/blocks/.eslintrc.cjs]: Found a workaround! I think the issue is related to how there are currently type-checking issues in Svelte's template sections. I was able to restore all rules by just passing references to functions that were typed in the script section, and that cleared everything up.

File: packages/blocks/src/lib/slam-map-2d/motion-path.svelte [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+      const y = Number.parseFloat(yString) / 1000;
+      points.push(x, y, 0);
+    }
+  }
+
+  const vertices = new Float32Array(points);
+  geometry.setPositions(vertices);
+};
+
+$: updatePath(path);
```
Comment (mcous) [inline packages/blocks/src/lib/slam-map-2d/motion-path.svelte:45]: Given that we create a new `LineGeometry` every time, would this make the intent more clear as a reactive derived variable?

```svelte
const pathStringToGeometry = (pathstr?: string): LineGeometry => {
  if (pathstr === undefined) {
    return new LineGeometry();
  }

  const points: number[]

  // ...

  const vertices = new Float32Array(points);
  const geometry = new LintGeometry()

  geometry.setPositions(vertices);

  return geometry
}

$: geometry = pathStringToGeometry(path)
```

File: packages/blocks/src/lib/slam-map-2d/motion-path.svelte [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+import { LineGeometry } from 'three/examples/jsm/lines/LineGeometry.js';
+import { renderOrder } from './constants';
+
+extend({ Line2, LineMaterial });
+
+export let path: string | undefined;
+
+let geometry: LineGeometry = new LineGeometry();
+
+const updatePath = (pathstr?: string) => {
```
Comment (mcous) [inline packages/blocks/src/lib/slam-map-2d/motion-path.svelte:23]: Should this function be extracted and unit tested?

File: packages/blocks/src/lib/slam-map-2d/motion-path.svelte [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+  const vertices = new Float32Array(points);
+  geometry.setPositions(vertices);
+};
+
+$: updatePath(path);
+
+</script>
+
+{#if path}
```
Comment (mcous) [inline packages/blocks/src/lib/slam-map-2d/motion-path.svelte:49]: Should this check `geometry` rather than `path`, given `geometry` is the variable that's actually used?

File: packages/blocks/src/lib/slam-map-2d/points.svelte [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+  material.size = size;
+}
+$: if (pointcloud) {
+  update(pointcloud);
+}
+$: raycaster.camera = camera.current as THREE.OrthographicCamera;
+
+onMount(() => dispatch('update', { center, radius }));
+
+raycaster.addEventListener('click', (event: THREE.Event) => {
```
Comment (mcous) [inline packages/blocks/src/lib/slam-map-2d/points.svelte:135]: Should this event listener be removed in `onDestroy`?

File: packages/blocks/src/lib/slam-map-2d/points.svelte
Comment (mcous) [inline packages/blocks/src/lib/slam-map-2d/points.svelte:1]: This component looks like a good candidate to split up into a relatively bare collaborator whose dependencies can be unit tested, given the amount of processing done in the `<script>` tag and the sparse template

File: packages/legacy/package.json
Comment (mcous) [inline packages/legacy/package.json:1]: Are the changes in `packages/legacy` necessary for this PR, or did they get added in the course of investigating the ESLint thing? If they are unrelated and can be dropped from this PR, that would be better given the number of `test.skips` that seem to have happened

**Review by mcous — APPROVED** [review]
LGTM! Some minor organizational comments (plus a `NaN` recommendation) and requests for whenever is convenient in terms of feature urgency

File: packages/blocks/src/lib/slam-map-2d/constants.ts
[the remark this answers is not in the document]
Comment (micheal-parks) [reply packages/blocks/src/lib/slam-map-2d/constants.ts:1]: Yeah totally with you concerning constants, helpers, utils. Renamed!

File: packages/blocks/src/lib/slam-map-2d/helpers.svelte
[the remark this answers is not in the document]
Comment (micheal-parks) [reply packages/blocks/src/lib/slam-map-2d/helpers.svelte:1]: In this - and only this - case I'd argue helpers is meaningful because of three's naming conventions here: https://threejs.org/docs/index.html#api/en/helpers/ArrowHelper

File: packages/blocks/src/lib/slam-map-2d/motion-path.svelte [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  geometry = new LineGeometry();
+
+  const points: number[] = [];
+
+  for (const xy of pathstr.split('\n')) {
+    const [xString, yString] = xy.split(',');
+    if (xString !== undefined && yString !== undefined) {
+      const x = Number.parseFloat(xString) / 1000;
+      const y = Number.parseFloat(yString) / 1000;
+      points.push(x, y, 0);
```
[the remark this answers is not in the document]
Comment (micheal-parks) [reply packages/blocks/src/lib/slam-map-2d/motion-path.svelte:40]: Done!

[the remark this answers is not in the document]
Comment (micheal-parks) [reply packages/blocks/src/lib/slam-map-2d/motion-path.svelte:40]: Done!

File: packages/blocks/src/lib/slam-map-2d/motion-path.svelte (hunk 5 above)
Comment (micheal-parks) [reply packages/blocks/src/lib/slam-map-2d/motion-path.svelte:23]: Yes! Just not yet - this is an advanced debugging feature that really only has internal use. There's a project in the works to hook the feature up to an API, at which point the incoming path data and therefore the function body is very likely to change. We should do it then if necessary (might just end up being plugging the response into the renderer rather than any parsing).

File: packages/blocks/src/lib/slam-map-2d/points.svelte (hunk 7 above)
Comment (micheal-parks) [reply packages/blocks/src/lib/slam-map-2d/points.svelte]: Good catch! Done.

File: packages/blocks/src/lib/slam-map-2d/points.svelte
Comment (micheal-parks) [reply packages/blocks/src/lib/slam-map-2d/points.svelte:1]: Good point, moved a majority of the logic into a separate .ts file

File: packages/blocks/package.json [hunk 9]
```diff
@@ -30,42 +30,54 @@
     "prime.css"
   ],
   "peerDependencies": {
```
[the remark this answers is not in the document]
Comment (micheal-parks) [reply packages/blocks/package.json:32]: I had the same thought - everything here is required for the navigation and slam blocks. Maybe something to re-assess as more blocks are added?

File: packages/blocks/src/lib/slam-map-2d/motion-path.svelte (hunk 4 above)
Comment (micheal-parks) [reply packages/blocks/src/lib/slam-map-2d/motion-path.svelte]: I like this, updated!

File: packages/legacy/package.json
Comment (micheal-parks) [reply packages/legacy/package.json:1]: Yeah, unfortunately I wasn't able to get CI to work until I upgraded every package. I'm not sure why and think my lack of experience with pnpm monorepos is probably to blame 😅 

All but two of the skipped tests are in a component we've completely stopped using due to event issues: VectorInput. The remaining two are testing v-code-editor, which we never ended up adopting in app.

File: packages/blocks/src/lib/slam-map-2d/helpers.svelte
[the remark this answers is not in the document]
Comment (mcous) [reply packages/blocks/src/lib/slam-map-2d/helpers.svelte:1]: Damn ok I'll raise my concerns with Three.js 🙃
